bugfix: `spack config edit` should work in a bad environment

tgamblin opened the pull request:
Fixes #2468. (this may have been fixed already but it's related)

This PR fixes a number of bugs (one per commit -- this should be rebased not squashed) to get to the main one:

1. `spack --debug` wasn't in the right place and wouldn't take effect for things at the beginning of `main()`.
2. Config errors had gotten really verbose -- they print a full stack trace and don't just show the user where the error is like they used to.
3. `spack config get <section>` and `spack config --scope <scope> edit <section>` weren't working in environments. `spack config get <section>`  should return the combined configuration for that section (including anything from `spack.yaml`), even in an environment, and `spack config edit` should respect the scope argument.
4. If you don't format `spack.yaml` correctly, `spack config edit` still fails and you have to edit your `spack.yaml` manually. Spack should be smarter than this.

- [x] Rework `_main()` and add some notes for maintainers on where things need to go for configuration to work properly.
- [x] Move config setup to *after* command-line parsing is done.
- [x] Add some code to `_main()` to defer `ConfigFormatError` when loading the environment, until we know what command is being run.  
- [x] Make `spack config edit` use `SPACK_ENV` instead of the config scope object to find `spack.yaml`, so it can work even if the environment is bad.
- [x] Reorder conditions in `cmd/config.py` to fix environment issues.
- [x] Tests

tgamblin committed:
errors: Rework error handling in `main()`

`main()` has grown, and in some cases code that can generate errors has gotten
outside the top-level try/catch in there. This means that simple errors like
config issues give you large stack traces, which shouldn't happen without
`--debug`.

- [x] Split `main()` into `main()` for the top-level error handling and
      `_main()` with all logic.

tgamblin committed:
config: ensure that options like `--debug` are set first

`spack --debug config edit` was not working properly -- it would not do show a
stack trace for configuration errors.

- [x] Rework `_main()` and add some notes for maintainers on where things need
      to go for configuration to work properly.
- [x] Move config setup to *after* command-line parsing is done.

Co-authored-by: scheibelp <scheibel1@llnl.gov>

tgamblin committed:
bugfix: `spack config get <section>` in environments

`spack config get <section>` was erroneously returning just the `spack.yaml`
for the environment.

It should return the combined configuration for that section (including
anything from `spack.yaml`), even in an environment.

- [x] reorder conditions in `cmd/config.py` to fix

scheibelp commented on lib/spack/spack/main.py:
```diff
@@ -782,36 +785,93 @@ def main(argv=None):
+            # print the context but delay this exception so that commands like
+            # `spack config edit` can still work with a bad environment.
+            e.print_context()
+            env_format_error = e
```
(minor: naming) I'm not sure why this is called `env_format_error` since it could be an issue with other config scopes (although it's true that this only occurs if we have an environment since the config is read earlier for that).

tgamblin replied:
I think this is fine to change to config_format_error or whatever

scheibelp replied:
From further experimentation it turns out this is only occurring when the `spack.yaml` has errors (so for example you can edit `packages.yaml` in the user scope, add errors, and still `spack config edit` it). So this variable name seems suitable.

scheibelp commented on lib/spack/spack/cmd/config.py:
```diff
@@ -143,17 +143,17 @@ def config_get(args):
 
-    if scope and scope.startswith('env:'):
+    if section is not None:
+        spack.config.config.print_section(section)
```
(I didn't think this would get posted: ignore it for now)

> `spack config get <section>` was erroneously returning just the `spack.yaml`
for the environment.

FWIW `spack config --scope=user get packages` seems to work for me (i.e. it prints everything up to the user scope and excludes the env scope). Furthermore `spack config get doesn't currently define what it's supposed to do for a specified scope (i.e. should it print just that scope, or should it print everything up to that scope?).

scheibelp commented on lib/spack/spack/main.py:
```diff
@@ -782,36 +785,93 @@ def main(argv=None):
+        try:
+            env = spack.cmd.find_environment(args)
+            if env:
+                ev.activate(env, args.use_env_repo, add_view=False)
```
(major) Currently getting an error:

```
# spack config --scope=defaults get packages
==> Error: activate() got an unexpected keyword argument 'add_view'
```

I think you can just remove `add_view`.

There appears to be an `environment.environment.activate` and an `environment.shell.activate`; only the latter accepts an `add_view` argument (but that isn't the one being used here I think.

scheibelp reviewed (changes requested):
I started looking through this in detail but noticed an error (perhaps things got out of sync) where I think most commands are broken: I prefixed that comment with "major".

scheibelp commented on share/spack/qa/run-shell-tests:
```diff
@@ -40,9 +40,11 @@ cd "$SPACK_ROOT"
 if [ "$COVERAGE" = true ]; then
     kcov "$SPACK_ROOT/coverage" "$QA_DIR/setup-env-test.sh" &> /dev/null
     kcov "$SPACK_ROOT/coverage" "$QA_DIR/completion-test.sh" &> /dev/null
+    kcov "$SPACK_ROOT/coverage" "$QA_DIR/test-env-cfg.sh" &> /dev/null
```
@tgamblin you mentioned adding coverage: AFAIK this is the way to do it, although the coverage numbers don't change at all (currently 43.18%). Did you have something else in mind for coverage?

tgamblin replied:
We use `kcov` for shell coverage, not python coverage -- and I think that makes sense for the other tests, but not for this one.  The other tests in here are supposed to cover `setup-env.sh` in various shells -- IIRC `bash` is the only one that actually does coverage properly (which is why there is only this if statement for the bash section) and `kcov` launches bash itself.  Anyway, that's not what we need here.

I think really `test-env-cfg.sh` should be called from `run-unit-tests.sh`, since it isn't actually testing shell functionality.

You should modify `test-env-cfg.sh` so that it invokes spack the way the unit test script does:

```console
$coverage_run $(which spack) ... args ...
```

`$coverage_run` is set by `setup.sh`, which all the test scripts use.  I could be wrong but I think it needs to be run from the unit test script to be included in the aggregation of coverage data -- I forget whether the shell one actually does that for `coverage.py` data or just `kcov` data.

scheibelp commented:
@spackbot re-run pipeline

scheibelp commented on share/spack/qa/test-env-cfg.sh:
```diff
@@ -0,0 +1,72 @@
+echo "Try 'spack config edit' with broken environment"
+manifest_path=`spack config edit --print-file`
+# Re-run command for coverage purposes
+$coverage_run spack config edit --print-file
```
I thought this is what you were suggesting in https://github.com/spack/spack/pull/24689#discussion_r737859729 but it doesn't affect the coverage numbers. It appears to have run, e.g. in https://github.com/spack/spack/runs/4028264529?check_suite_focus=true: 

```
...
====== 3088 passed, 107 skipped, 7 xfailed, 4 xpassed in 1954.52 seconds =======
==> Updating view at /home/runner/work/spack/spack/var/spack/environments/broken-cfg-env/.spack-env/view
==> Created environment 'broken-cfg-env' in /home/runner/work/spack/spack/var/spack/environments/broken-cfg-env
...
```

tgamblin replied:
It looks like `run-unit-tests` sources `setup.sh`, but because it doesn't export the variables that get set there (like `coverage_run`), they aren't making it to your script.  I was able to determine this locally.

I think you can fix it by just sourcing setup.sh in your script.

scheibelp replied:
OK excellent with that I have bumped the coverage from 43% to 59%. The details on codecov do not attribute this to `test-env-cfg.sh` (i.e. that is ignored) but instead to handling of `spack --pdb...`.

If it is agreeable not to test that, I'd say the current coverage is sufficient.

tgamblin replied:
I looked at the diff on `codecov`, and it looks like the coverage numbers are off by one (+1) from the actual lines they're supposed to be covering.  I don't think we should worry about that too much right now, just saying.

The lines we care about for *this specific PR* are covered so I think we can call this good.  The rest of `main()` was poorly covered to begin with, but I think with your script it can be more covered in the near future.  I say let's save renaming `test-env-cfg.sh` to `test-main.sh` for some future PR and get all the corner cases covered then.

If you're ok with this plan go ahead and approve and let's merge this.

tgamblin committed:
bugfix: config edit should work with a malformed `spack.yaml`

If you don't format `spack.yaml` correctly, `spack config edit` still fails and
you have to edit your `spack.yaml` manually.

- [x] Add some code to `_main()` to defer `ConfigFormatError` when loading the
  environment, until we know what command is being run.

- [x] Make `spack config edit` use `SPACK_ENV` instead of the config scope
  object to find `spack.yaml`, so it can work even if the environment is bad.

Co-authored-by: scheibelp <scheibel1@llnl.gov>

adamjstewart commented:
@tgamblin @scheibelp the first commit of this PR broke our style tests.

### Before

```console
$ git checkout 732be7dec6
$ spack style --no-isort --no-mypy
==> Running style checks on spack
  selected: flake8
==> Modified files
  ../../../var/spack/repos/builtin/packages/hdf5/package.py
==> Running flake8 checks
  ../../../var/spack/repos/builtin/packages/hdf5/package.py:30: [E303] too many blank lines (2)
  flake8 found errors
==> Error: spack style found errors
$ echo $?
1
```

### After

```console
$ git checkout 56ad721eb5
$ spack style --no-isort --no-mypy
==> Running style checks on spack
  selected: flake8
==> Modified files
  ../../../var/spack/repos/builtin/packages/hdf5/package.py
==> Running flake8 checks
  ../../../var/spack/repos/builtin/packages/hdf5/package.py:30: [E303] too many blank lines (2)
  flake8 found errors
==> Error: spack style found errors
$ echo $?
0
```
I noticed a PR where style should have failed but the tests pass just fine: https://github.com/spack/spack/runs/4049232114

scheibelp commented:
Thanks @adamjstewart, https://github.com/spack/spack/pull/27077 should address this
